### PR TITLE
Handle direct CanncelationException on timeout in JdkClientHttpRequest

### LINKED ISSUE #33973: Random CancellationException instead of ResourceAccessException after upgrading to Spring Boot 3.4.0
Hi Spring-Team,

After upgrading to Spring Boot 3.4.0, we have encountered an issue with RestTemplate where random CancellationExceptions are thrown instead of the expected ResourceAccessException during timeout scenarios.

Observations:
This behavior is inconsistent; sometimes a CancellationException is thrown, and other times a ResourceAccessException is thrown.
This issue did not occur in previous versions of Spring Boot.

Reproduction:
We have created a test case to demonstrate the issue: https://github.com/SimonPNorra/timeout-execption-case

Steps to Reproduce:
Run the test case included in the project and observe the exceptions thrown. You will notice either a CancellationException or a ResourceAccessException occurring randomly.

Expected Behavior:
Consistent behavior with ResourceAccessException thrown during timeout scenarios.

Actual Behavior:
Random exceptions are thrown: either CancellationException or ResourceAccessException.

Please let us know if further information is required.

Thank you.

Regards

### PR DESCRIPTION
**Pull request by giampa91**
CancellationExceptions are thrown instead of the expected ResourceAccessException during timeout scenarios. 
Handle the CancellationExceptions in order to throw an ResourceAccessException when timeout occurred

Closes gh-33973

### CODE REVIEW

File: spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java [hunk 1]
```diff
@@ -96,12 +96,13 @@ public URI getURI() {
 	@Override
 	protected ClientHttpResponse executeInternal(HttpHeaders headers, @Nullable Body body) throws IOException {
 		CompletableFuture<HttpResponse<InputStream>> responseFuture = null;
+		TimeoutHandler timeoutHandler = null;
```
Comment (rstoyanchev) [inline spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java:100]: There seems to be no purpose to this change that I can see. Is this intended?

File: spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java [hunk 2]
```diff
@@ -224,6 +228,7 @@ public ByteBuffer map(byte[] b, int off, int len) {
 	private static final class TimeoutHandler {
 
 		private final CompletableFuture<Void> timeoutFuture;
+		private boolean isTimeout=false;
```
Comment (rstoyanchev) [inline spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java]: Likewise here, this flag is never used.

**Review by rstoyanchev — CHANGES_REQUESTED** [review]
There seems to be no purpose to moving the TimeoutHandler

File: spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java (hunk 2 above)
Comment (giampa91) [reply spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java:231]: the flag is used in line 240 and is setted to true, and on default is false

Comment (giampa91) [reply spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java]: sorry my mistake, now this variable has a purpose

File: spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java (hunk 1 above)
Comment (giampa91) [reply spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java:100]: this change was to make the TimeoutHandler readable outside the try

**Comment by giampa91** [comment]
Thank you for your patience @rstoyanchev  ! After a careful review, I realized I hadn’t double-checked the code thoroughly enough.

I’ve now updated the implementation to better distinguish between a cancelled request and a genuine timeout by explicitly checking the timeout variable.

Additionally, I’ve added tests covering both scenarios. While there might still be room for optimization, I believe this approach reflects real-world usage and edge cases.

**Review by rstoyanchev — APPROVED** [review]
Okay the intent is more clear now.
